### PR TITLE
Improve torch.compile device mismatch errors

### LINKED ISSUE #159133: Inconsistent Error Message for Cross-Device Input in torch.compile
## Repro Script

```python
import torch
import torch.nn as nn

class BuggyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(10, 20)

    def forward(self, x):
        return self.fc(x)

def minimal_repro():
    model = BuggyModel().cuda().eval()  # Model on CUDA
    x = torch.randn(1, 10)              # Input on CPU

    model(x.to('cuda'))

    try:
        model(x)
    except Exception as e:
        print(f"\n[BUG_TRIG] Eager mode failed: {e}")

    try:
        cmodel = torch.compile(model)
        cmodel(x)  # CPU input, CUDA model
    except Exception as e:
        print(f"\n[BUG_TRIG] Compile failed: {e}")

if __name__ == "__main__":
    minimal_repro()

```

## Output

```
[BUG_TRIG] Eager mode failed: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat1 in method wrapper_CUDA_addmm)

[BUG_TRIG] Compile failed: Dynamo failed to run FX node with fake tensors: call_function <built-in function linear>(*(FakeTensor(..., size=(1, 10)), Parameter(FakeTensor(..., device='cuda:0', size=(20, 10), requires_grad=True)), Parameter(FakeTensor(..., device='cuda:0', size=(20,), requires_grad=True))), **{}): got RuntimeError('Unhandled FakeTensor Device Propagation for aten.mm.default, found two different devices cpu, cuda:0')

from user code:
   File ".../test_coverage/for_test_1.py", line 10, in forward
    return self.fc(x)
  File ".../torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
```

## Problem
In eager mode, the user gets a clear error:
```
Expected all tensors to be on the same device...
```
But in `torch.compile`, the error is instead:
```
Unhandled FakeTensor Device Propagation for aten.mm.default...
```

This message:

+ Comes from the FakeTensor system,

+ Might confuse users into thinking this is a bug in Dynamo or FX tracing logic itself, rather than the real problem being a device mismatch.

## Question
Would it be possible for torch.compile to:

+ Detect this common cross-device situation earlier,

+ Or provide an equally clear and actionable error message as in eager mode?

Right now the trace suggests internal problems ("Unhandled FakeTensor propagation") that aren't very informative from a user perspective.

### Versions

```
PyTorch version: 2.7.1a0+gite2d141d
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

cc @malfet @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @Lucaskabela

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185412

FakeTensor's device propagation raised a generic RuntimeError when non-scalar fake tensors came from different devices. Dynamo wrapped that through the generic fake tensor FX-node failure path, so a user CPU input with CUDA parameters reported "Unhandled FakeTensor Device Propagation" instead of the actionable same-device error eager mode gives.

This introduces a specific FakeTensorDeviceMismatchError for that condition and teaches Dynamo to map it to a Tensor device mismatch graph break with direct guidance to place inputs, parameters, and buffers on the same device. I considered translating generic FakeTensor RuntimeErrors in Dynamo, but a dedicated exception keeps real FakeTensor propagation bugs on the existing internal-error path.

Fixes #159133
Generated by my agent

Test Plan:
- python test/dynamo/test_error_messages.py ErrorMessagesTest.test_fx_node_error_cross_device ErrorMessagesTest.test_fx_node_error_bad_user_code
- python test/test_fake_tensor.py FakeTensorTest.test_op_with_zero_dim_bypassed FakeTensorTest.test_add_one_dim_single_elem_cpu_with_cuda_tensor FakeTensorPreferDeviceType.test_fake_tensor_prefer_device_type
- lintrunner -a
- git diff --cached --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the assertion typo noted by Claude, added the release notes: dynamo label, marked the PR ready, requested reviewers, and resubmitted on current main. The prior Dr. CI failures were SIGKILLs in unrelated test_serialization_2gb_file jobs; CI is now rerunning on the updated head.

**Review by karthickai — APPROVED** [review]
